cirq-google/cirq_google/workflow/quantum_executable.py

```diff
@@ -53,6 +53,10 @@ class KeyValueExecutableSpec(ExecutableSpec):
     executable_family: str
     key_value_pairs: Tuple[Tuple[str, Any], ...] = ()
 
+    @classmethod
+    def _json_cirq_type_(cls) -> str:
+        return 'cirq.google.KeyValueExecutableSpec'
```

Comment:
slightly confused as to why these are not cirq_google.  This seems to be the reason you need these to override the default behavior?

Reply:
Two things:

1. These are `cirq.google.foo` because that's what appears in [`json_resolver_cache.py`](https://github.com/quantumlib/Cirq/blob/1d7436ff22891e43978572c1bff78cbac6cd0bef/cirq-google/cirq_google/json_resolver_cache.py#L44-L51). Changing to cirq_google would require a backwards-incompatible change to our serialization format.
2. The default is no module name (e.g. `KeyValueExecutableSpec`). Even if this was `cirq_google`, I'd still need the override for these cases.

Reply:
Up until now, the source of truth for the cirq_type field in the resultant dictionary and/or serialized form has been from the value returned in `_json_dict_`.

1. why do we need to query this separate from the value there
2. now this information is redundant and could become out-of-sync

There's also structure we're missing here: the cirq_type is still the class name, just namespaced. Namespaces are one honking great idea.

Reply:
1. `_json_dict_` is an instance method - I can't call it to retrieve the `cirq_type` field without instantiating the type, which is what type-serialization is trying to avoid.
2. True. If you'd like, I could modify the affected classes to use their `json_cirq_type` method in constructing their `_json_dict_`.

Namespacing comes back to the backwards-compatibility concern I noted above: all old serialization formats **must** be deserializable. If we want to globally enforce namespaces in `cirq_type` values, pre-v1.0 is a good time to do it, but that should be raised as a separate issue.

Reply:
The "cirq_type" value used to be a completely arbitrary string with the only condition that each object type provide a unique one. It was completely specified by the specific key-value pair in the dictionary returned by `_json_dict_` and was only ever consumed by the resolver. By _convention_, it was the name of the class for top-level cirq objects. This convention was aided by the `obj_to_dict_helper` and `dataclass_json_dict` helper functions that could help one implement the `_json_dict_` method. There was also a convention I imagined where each objects' "cirq_type" would be namespaced with dots (`.`) like `cirq.google.Thing`. That's the convention aided by the `namespace` argument in `obj_to_dict_helper` and `dataclass_json_dict`.

Philosophically, I think this PR represents a change from "by convention it works this way" to "by design it works this way". I think this is ok, but now our protocols (`_json_dict_` and `_json_cirq_type_`) have this redundancy. What do we do if someone defines a class where these conflict? We can modify some or all of the classes under our control to use `_json_cirq_type_` to define `_json_dict_` but that doesn't stop a future developer or a developer piggy-backing on the cirq json infrastructure (this happens!) from mucking it up.

Finally, re: namespace. I think you misunderstood. I don't care about "cirq.google" vs "cirq_google". I want the json_cirq_type protocol to have support for "this class's json_cirq_type still follow the rules: it's the python class name; i've just prepended some disambiguating string with a dot to the front of it".

What would happen if I did:

```python
    @classmethod
    def _json_cirq_type_(cls) -> str:
        return 'cirq.google.KeyValueExecuatbleSpec'
```

did you notice the typo? Compare with e.g. 

```python
class Thing:
  cirq_type_namespace = 'recirq.qaoa'
  
  def __init__(self, x, y):
     ...

  ...

def default_json_cirq_type:
  if hasattr(obj, 'json_cirq_type'):
    return obj.json_cirq_type()
  
  if hasattr(obj, 'cirq_type_namespace'):
     return obj.cirq_type_namespace + obj.__class__.__name__
  
  return obj.__class__.__name__
```

Reply:
> Philosophically, I think this PR represents a change from "by convention it works this way" to "by design it works this way". I think this is ok, but now our protocols (`_json_dict_` and `_json_cirq_type_`) have this redundancy. What do we do if someone defines a class where these conflict?

The round-trip serialization test protects against conflicts in these protocols: both must match the value used in `json_resolver_cache`, or deserialization will fail. (The typo example above would encounter this test failure.) I see what you mean about redundancy, but it's also present in the suggested "namespaces" design: even with `default_json_cirq_type`, the `Thing._json_dict_` method would need to refer to `cirq_type_namespace` to be consistent.

**What is the main goal here?** If we just want to support namespacing, I can do that in this PR. However, if the goal is to remove redundancy, we should lean into the new design and have the `cirq_type` field (both namespace and class name) be automatically populated by `cirq.to_json`. That would be a more complicated change.

Reply:
> If we just want to support namespacing, I can do that in this PR. However, if the goal is to remove redundancy, we should lean into the new design and have the cirq_type field (both namespace and class name) be automatically populated by cirq.to_json. That would be a more complicated change.

I agree with this analysis. My philosophical point is that: while this PR looks innocuous enough, the complete implementation of this new protocol would be to lean into the new design and do the more complicated change.

I would appreciate it if you support namespacing in this PR and open a follow-on issue for removing the redundancy

Reply:
Opened #4698 to track the redundancy-removal step.

Reply:
Added namespacing.